File: solutions/erlang/bank-account/1/src/bank_account.erl
%% Bank Account module using gen_server behaviour.
-module(bank_account).
-behaviour(gen_server).

%% Exporting client functions and server callbacks
-export([balance/1, charge/2, close/1, create/0, deposit/2, withdraw/2]).
-export([init/1, handle_call/3, handle_cast/2]).

%% Record definition for the account, with default state 'open' and balance of 0.
-record(account, {state = open, balance = 0}).

%% Client API functions

%% Creates a new bank account (starts the gen_server process).
create() ->
    {ok, Pid} = gen_server:start_link(?MODULE, [], []),
    Pid.

%% Returns the balance of the account by making a synchronous call to the server.
balance(Pid) ->
    gen_server:call(Pid, balance).

%% Charges a specified amount from the account (synchronous call).
charge(Pid, Amount) ->
    gen_server:call(Pid, {charge, Amount}).

%% Closes the bank account (synchronous call).
close(Pid) ->
    gen_server:call(Pid, close).

%% Deposits a specified amount into the account (asynchronous call).
deposit(Pid, Amount) ->
    gen_server:cast(Pid, {deposit, Amount}).

%% Withdraws a specified amount from the account (synchronous call).
withdraw(Pid, Amount) ->
    gen_server:call(Pid, {withdraw, Amount}).

%% Server Callbacks

%% Initializes the server with an account record.
init([]) ->
    {ok, #account{}}.

%% Handle synchronous (call) messages

%% If the account is closed, return an error.
handle_call(_, _, Account) when Account#account.state == closed ->
    {reply, {error, account_closed}, Account};

%% Handle closing the account. Return the current balance and set state to closed.
handle_call(close, _From, Account) ->
    {reply, Account#account.balance, Account#account{state = closed}};

%% Handle balance inquiry. Return the current balance.
handle_call(balance, _From, Account) ->
    {reply, Account#account.balance, Account};

%% Handle withdrawal requests.
handle_call({withdraw, Amount}, _From, Account) ->
    Balance = Account#account.balance,
    if
        Amount < 0 ->  %% Reject negative withdrawals.
            {reply, 0, Account};
        Amount > Balance ->  %% If withdrawal exceeds balance, withdraw all funds.
            {reply, Balance, Account#account{balance = 0}};
        true ->  %% Otherwise, deduct the amount from the balance.
            {reply, Amount, Account#account{balance = Balance - Amount}}
    end;

%% Handle charge requests (similar to withdrawal).
handle_call({charge, Amount}, _From, Account) ->
    Balance = Account#account.balance,
    if
        Amount < 0 ->  %% Reject negative charges.
            {reply, 0, Account};
        Amount > Balance ->  %% Reject if charge exceeds balance.
            {reply, 0, Account};
        true ->  %% Deduct the charge from the balance.
            {reply, Amount, Account#account{balance = Balance - Amount}}
    end.

%% Handle asynchronous (cast) messages

%% If the account is closed, ignore the message.
handle_cast(_, Account) when Account#account.state == closed ->
    {noreply, Account};

%% Handle deposit requests.
handle_cast({deposit, Amount}, Account) ->
    if
        Amount < 0 ->  %% Ignore negative deposits.
            {noreply, Account};
        true ->  %% Add the deposit to the current balance.
            {noreply, Account#account{balance = Account#account.balance + Amount}}
    end.

File: solutions/javascript/bank-account/1/bank-account.js
/**
 * Represents a bank account with basic operations such as opening, closing,
 * depositing, and withdrawing funds.
 */
export class BankAccount {
  /**
   * Opens the bank account. Throws an error if the account is already open.
   */
  open() {
    if (this.isOpen) {
      throw new ValueError('Account is already open.');
    }
    this._open = true;
    this._balance = 0;
  }

  /**
   * Closes the bank account. Throws an error if the account is not open.
   */
  close() {
    this.verifyAccount();
    this._open = false;
  }

  /**
   * Verifies that the account is open. Throws an error if the account is not open.
   * @private
   */
  verifyAccount() {
    if (!this.isOpen) {
      throw new ValueError('Account is not open.');
    }
  }

  /**
   * Deposits the specified amount into the account. Throws an error if the amount
   * is negative or if the account is not open.
   * @param {number} amount - The amount to deposit.
   */
  deposit(amount) {
    this.verifyAccount();
    this.verifyDeposit(amount);
    this._balance += amount;
  }

  /**
   * Verifies that the deposit amount is valid. Throws an error if the amount
   * is negative.
   * @param {number} amount - The amount to verify.
   * @private
   */
  verifyDeposit(amount) {
    if (amount < 0) {
      throw new ValueError('Deposit amount must be non-negative.');
    }
  }

  /**
   * Withdraws the specified amount from the account. Throws an error if the amount
   * is negative, exceeds the current balance, or if the account is not open.
   * @param {number} amount - The amount to withdraw.
   */
  withdraw(amount) {
    this.verifyAccount();
    this.verifyWithdraw(amount);
    this._balance -= amount;
  }

  /**
   * Verifies that the withdrawal amount is valid. Throws an error if the amount
   * is negative or exceeds the current balance.
   * @param {number} amount - The amount to verify.
   * @private
   */
  verifyWithdraw(amount) {
    if (amount < 0 || amount > this._balance) {
      throw new ValueError('Invalid withdrawal amount.');
    }
  }

  /**
   * Gets the current balance of the account. Throws an error if the account is not open.
   * @returns {number} The current balance.
   */
  get balance() {
    this.verifyAccount();
    return this._balance;
  }

  /**
   * Checks if the account is open.
   * @returns {boolean} True if the account is open, false otherwise.
   */
  get isOpen() {
    return this._open;
  }
}

/**
 * Represents a value error, typically used for invalid operations or inputs.
 */
export class ValueError extends Error {}